etna/packages/etna-py/src/mountetna/metis.py

```diff
@@ -504,19 +519,7 @@ def upload_parts(
                 )
 
                 unsent_zero_byte_file = False
-            except RequestException as e:
```

Comment:
Seems like when an exception is raised inside a generator (`yield` / `yield from`), it doesn't get handled inside the generator itself, but the calling method has to catch and handle it. You can see this in the stacktrace below, where the exception bubbles up from line 517 in `metis.py`, which is the generator call inside of this `except` block .. and which isn't subsequently caught.

```
[2022-08-01, 14:32:15 PDT] {remote_helpers_base.py:133} INFO - Uploading blob... (9.24%)
[2022-08-01, 14:32:17 PDT] {logging_mixin.py:109} INFO - Error 502, <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>502 Bad Gateway</title>
</head><body>
<h1>Bad Gateway</h1>
<p>The proxy server received an invalid
response from an upstream server.<br />
</p>
</body></html>
. Retrying...
[2022-08-01, 14:32:47 PDT] {taskinstance.py:1718} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/mountetna/metis.py", line 495, in upload_parts
    self.upload_blob(
  File "/home/airflow/.local/lib/python3.8/site-packages/mountetna/metis.py", line 439, in upload_blob
    response = self.session.post(
  File "/home/airflow/.local/lib/python3.8/site-packages/requests/sessions.py", line 577, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/requests/sessions.py", line 529, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/requests/sessions.py", line 652, in send
    r = dispatch_hook('response', hooks, r, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/requests/hooks.py", line 31, in dispatch_hook
    _hook_data = hook(hook_data, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/mountetna/etna_base.py", line 38, in assert_status
    raise HTTPError(error_message, response=response)
requests.exceptions.HTTPError: Bad Gateway

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1334, in _run_raw_task
    self._execute_task_with_callbacks(context)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1460, in _execute_task_with_callbacks
    result = self._execute_task(context, self.task)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1516, in _execute_task
    result = execute_callable(context=context)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/decorators/base.py", line 134, in execute
    return_value = super().execute(context)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/operators/python.py", line 174, in execute
    return_value = self.execute_callable()
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/operators/python.py", line 188, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/etna/etls/cat.py", line 162, in ingest
    self.handle_metis_ingest(
  File "/home/airflow/.local/lib/python3.8/site-packages/etna/etls/remote_helpers_base.py", line 123, in handle_metis_ingest
    for blob in metis.upload_file(
  File "/home/airflow/.local/lib/python3.8/site-packages/mountetna/metis.py", line 517, in upload_parts
    yield from self.upload_parts(
  File "/home/airflow/.local/lib/python3.8/site-packages/mountetna/metis.py", line 491, in upload_parts
    blob_data = upload.read_next_blob()
  File "/home/airflow/.local/lib/python3.8/site-packages/mountetna/metis.py", line 288, in read_next_blob
    raise ValueError(
ValueError: Upload source stream requires restart, upload has failed.
```

So, moving the retry logic to `upload_file`, to see if the upload will retry more than once.

Reply:
Hold up, this seems wrong, let me check.

Reply:
:+1: Was hard to test without triggering an exception somehow ... and now that production is stable, haven't seen a retry yet ...

Reply:
Oh oh, is it because `ValueError` is not caught by the retry block, so it just gets re-raised and bubbles up? Hm....

Reply:
IE: Make it possible for us to provide alternative functions rather than upload_start and upload_blob to the retry logic, making it much simpler to test

Reply:
:+1:

Reply:
```
[2022-08-02, 10:59:10 PDT] {remote_helpers_base.py:133} INFO - Uploading blob... (77.69%)
[2022-08-02, 10:59:13 PDT] {taskinstance.py:1718} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1334, in _run_raw_task
    self._execute_task_with_callbacks(context)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1460, in _execute_task_with_callbacks
    result = self._execute_task(context, self.task)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/models/taskinstance.py", line 1516, in _execute_task
    result = execute_callable(context=context)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/decorators/base.py", line 134, in execute
    return_value = super().execute(context)
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/operators/python.py", line 174, in execute
    return_value = self.execute_callable()
  File "/home/airflow/.local/lib/python3.8/site-packages/airflow/operators/python.py", line 188, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/etna/etls/cat.py", line 162, in ingest
    self.handle_metis_ingest(
  File "/home/airflow/.local/lib/python3.8/site-packages/etna/etls/remote_helpers_base.py", line 123, in handle_metis_ingest
    for blob in metis.upload_file(
  File "/home/airflow/.local/lib/python3.8/site-packages/mountetna/metis.py", line 523, in upload_parts
    raise e
  File "/home/airflow/.local/lib/python3.8/site-packages/mountetna/metis.py", line 510, in upload_parts
    self.upload_blob(
  File "/home/airflow/.local/lib/python3.8/site-packages/mountetna/metis.py", line 439, in upload_blob
    response = self.session.post(
  File "/home/airflow/.local/lib/python3.8/site-packages/requests/sessions.py", line 577, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/requests/sessions.py", line 529, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/requests/sessions.py", line 652, in send
    r = dispatch_hook('response', hooks, r, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/requests/hooks.py", line 31, in dispatch_hook
    _hook_data = hook(hook_data, **kwargs)
  File "/home/airflow/.local/lib/python3.8/site-packages/mountetna/etna_base.py", line 38, in assert_status
    raise HTTPError(error_message, response=response)
requests.exceptions.HTTPError: Service Unavailable
```

Yeah, I'm thinking the retry logic definitely isn't working as I thought. Will have to better understand generators and how they interact with exceptions, because seems like I'm not wrapping it correctly.